File: spline.js
// (c) Copyright 2017, Sean Connelly (@voidqk), http://syntheti.cc
// MIT License
// Project Home: https://github.com/voidqk/spline

// Catmull-Rom spline implementation (2d, 3d, 4d)

// `p0`, `p1`, `p2`, `p3` are points in format of [x, y]
// `tension` is number between 0 and 1 (0.5 is good default)
// `segs` is the number of segments to return
// returns a list of points of length `segs`, where `result[0]` is `p1`, and `result[segs - 1]` is
// approaching `p2`
function spline2d_segment(p0, p1, p2, p3, tension, segs){
	var p0x = p0[0], p0y = p0[1];
	var p1x = p1[0], p1y = p1[1];
	var p2x = p2[0], p2y = p2[1];
	var p3x = p3[0], p3y = p3[1];

	function tb(ax, ay, bx, by){
		var dx = bx - ax;
		var dy = by - ay;
		return Math.max(Math.pow(dx * dx + dy * dy, tension * 0.5), 0.0001);
	}
	var t1 = tb(p0x, p0y, p1x, p1y);
	var t2 = tb(p1x, p1y, p2x, p2y) + t1;
	var t3 = tb(p2x, p2y, p3x, p3y) + t2;

	var dt21 = t2 - t1;
	var dt31 = t3 - t1;
	var dt32 = t3 - t2;

	var res = [];
	for (var s = 0; s < segs; s++){
		var t = dt21 * s / segs + t1;

		var dt1 = t1 - t;
		var dt2 = t2 - t;
		var dt3 = t3 - t;

		var A1x = (dt1 * p0x + t * p1x) / t1;
		var A1y = (dt1 * p0y + t * p1y) / t1;

		var A2x = (dt2 * p1x - dt1 * p2x) / dt21;
		var A2y = (dt2 * p1y - dt1 * p2y) / dt21;

		var A3x = (dt3 * p2x - dt2 * p3x) / dt32;
		var A3y = (dt3 * p2y - dt2 * p3y) / dt32;

		var B1x = (dt2 * A1x + t * A2x) / t2;
		var B1y = (dt2 * A1y + t * A2y) / t2;

		var B2x = (dt3 * A2x - dt1 * A3x) / dt31;
		var B2y = (dt3 * A2y - dt1 * A3y) / dt31;

		res.push([
			(dt2 * B1x - dt1 * B2x) / dt21,
			(dt2 * B1y - dt1 * B2y) / dt21
		]);
	}
	return res;
}

// `p0`, `p1`, `p2`, `p3` are points in format of [x, y, z]
// `tension` is number between 0 and 1 (0.5 is good default)
// `segs` is the number of segments to return
// returns a list of points of length `segs`, where `result[0]` is `p1`, and `result[segs - 1]` is
// approaching `p2`
function spline3d_segment(p0, p1, p2, p3, tension, segs){
	var p0x = p0[0], p0y = p0[1], p0z = p0[2];
	var p1x = p1[0], p1y = p1[1], p1z = p1[2];
	var p2x = p2[0], p2y = p2[1], p2z = p2[2];
	var p3x = p3[0], p3y = p3[1], p3z = p3[2];

	function tb(ax, ay, az, bx, by, bz){
		var dx = bx - ax;
		var dy = by - ay;
		var dz = bz - az;
		return Math.max(Math.pow(dx * dx + dy * dy + dz * dz, tension * 0.5), 0.0001);
	}
	var t1 = tb(p0x, p0y, p0z, p1x, p1y, p1z);
	var t2 = tb(p1x, p1y, p1z, p2x, p2y, p2z) + t1;
	var t3 = tb(p2x, p2y, p2z, p3x, p3y, p3z) + t2;

	var dt21 = t2 - t1;
	var dt31 = t3 - t1;
	var dt32 = t3 - t2;

	var res = [];
	for (var s = 0; s < segs; s++){
		var t = dt21 * s / segs + t1;

		var dt1 = t1 - t;
		var dt2 = t2 - t;
		var dt3 = t3 - t;

		var A1x = (dt1 * p0x + t * p1x) / t1;
		var A1y = (dt1 * p0y + t * p1y) / t1;
		var A1z = (dt1 * p0z + t * p1z) / t1;

		var A2x = (dt2 * p1x - dt1 * p2x) / dt21;
		var A2y = (dt2 * p1y - dt1 * p2y) / dt21;
		var A2z = (dt2 * p1z - dt1 * p2z) / dt21;

		var A3x = (dt3 * p2x - dt2 * p3x) / dt32;
		var A3y = (dt3 * p2y - dt2 * p3y) / dt32;
		var A3z = (dt3 * p2z - dt2 * p3z) / dt32;

		var B1x = (dt2 * A1x + t * A2x) / t2;
		var B1y = (dt2 * A1y + t * A2y) / t2;
		var B1z = (dt2 * A1z + t * A2z) / t2;

		var B2x = (dt3 * A2x - dt1 * A3x) / dt31;
		var B2y = (dt3 * A2y - dt1 * A3y) / dt31;
		var B2z = (dt3 * A2z - dt1 * A3z) / dt31;

		res.push([
			(dt2 * B1x - dt1 * B2x) / dt21,
			(dt2 * B1y - dt1 * B2y) / dt21,
			(dt2 * B1z - dt1 * B2z) / dt21
		]);
	}
	return res;
}

// `p0`, `p1`, `p2`, `p3` are points in format of [x, y, z, w]
// `tension` is number between 0 and 1 (0.5 is good default)
// `segs` is the number of segments to return
// returns a list of points of length `segs`, where `result[0]` is `p1`, and `result[segs - 1]` is
// approaching `p2`
function spline4d_segment(p0, p1, p2, p3, tension, segs){
	var p0x = p0[0], p0y = p0[1], p0z = p0[2], p0w = p0[3];
	var p1x = p1[0], p1y = p1[1], p1z = p1[2], p1w = p1[3];
	var p2x = p2[0], p2y = p2[1], p2z = p2[2], p2w = p2[3];
	var p3x = p3[0], p3y = p3[1], p3z = p3[2], p3w = p3[3];

	function tb(ax, ay, az, aw, bx, by, bz, bw){
		var dx = bx - ax;
		var dy = by - ay;
		var dz = bz - az;
		var dw = bw - aw;
		return Math.max(Math.pow(dx * dx + dy * dy + dz * dz + dw * dw, tension * 0.5), 0.0001);
	}
	var t1 = tb(p0x, p0y, p0z, p0w, p1x, p1y, p1z, p1w);
	var t2 = tb(p1x, p1y, p1z, p1w, p2x, p2y, p2z, p2w) + t1;
	var t3 = tb(p2x, p2y, p2z, p2w, p3x, p3y, p3z, p3w) + t2;

	var dt21 = t2 - t1;
	var dt31 = t3 - t1;
	var dt32 = t3 - t2;

	var res = [];
	for (var s = 0; s < segs; s++){
		var t = dt21 * s / segs + t1;

		var dt1 = t1 - t;
		var dt2 = t2 - t;
		var dt3 = t3 - t;

		var A1x = (dt1 * p0x + t * p1x) / t1;
		var A1y = (dt1 * p0y + t * p1y) / t1;
		var A1z = (dt1 * p0z + t * p1z) / t1;
		var A1w = (dt1 * p0w + t * p1w) / t1;

		var A2x = (dt2 * p1x - dt1 * p2x) / dt21;
		var A2y = (dt2 * p1y - dt1 * p2y) / dt21;
		var A2z = (dt2 * p1z - dt1 * p2z) / dt21;
		var A2w = (dt2 * p1w - dt1 * p2w) / dt21;

		var A3x = (dt3 * p2x - dt2 * p3x) / dt32;
		var A3y = (dt3 * p2y - dt2 * p3y) / dt32;
		var A3z = (dt3 * p2z - dt2 * p3z) / dt32;
		var A3w = (dt3 * p2w - dt2 * p3w) / dt32;

		var B1x = (dt2 * A1x + t * A2x) / t2;
		var B1y = (dt2 * A1y + t * A2y) / t2;
		var B1z = (dt2 * A1z + t * A2z) / t2;
		var B1w = (dt2 * A1w + t * A2w) / t2;

		var B2x = (dt3 * A2x - dt1 * A3x) / dt31;
		var B2y = (dt3 * A2y - dt1 * A3y) / dt31;
		var B2z = (dt3 * A2z - dt1 * A3z) / dt31;
		var B2w = (dt3 * A2w - dt1 * A3w) / dt31;

		res.push([
			(dt2 * B1x - dt1 * B2x) / dt21,
			(dt2 * B1y - dt1 * B2y) / dt21,
			(dt2 * B1z - dt1 * B2z) / dt21,
			(dt2 * B1w - dt1 * B2w) / dt21
		]);
	}
	return res;
}

//
// convenience functions for generating splines using a list of points
//

var spline2d, spline3d, spline4d;
(function(){
	function makefunc(spline){
		// `points` is a list of points with the proper dimensions [[x, y], ...] or
		// [[x, y, z], ...] or [[x, y, z, w], ...]
		// `tension` is a number between 0 and 1 (0.5 is good default)
		// `segs` is the number of segments generated between two adjacent points
		// `close` is a boolean that indicates whether the returned curve should be a closed loop
		return function(points, tension, segs, close){
			var P = points;
			var L = P.length;

			if (L <= 0)
				return [];
			else if (L == 1)
				return [P[0].concat()];
			else if (L == 2){
				if (close){
					return spline(P[1], P[0], P[1], P[0], tension, segs).concat(
						spline(P[0], P[1], P[0], P[1], tension, segs));
				}
				return spline(P[0], P[0], P[1], P[1], tension, segs).concat([P[1].concat()]);
			}

			var res;
			if (close){
				res = spline(P[L - 1], P[0], P[1], P[2], tension, segs);
				if (L == 3){
					return res.concat(
						spline(P[0], P[1], P[2], P[0], tension, segs)).concat(
						spline(P[1], P[2], P[0], P[1], tension, segs));
				}
			}
			else{
				res = spline(P[0], P[0], P[1], P[2], tension, segs);
				if (L == 3){
					return res.concat(spline(P[0], P[1], P[2], P[2], tension, segs)).concat(
						[P[0].concat()]);
				}
			}

			for (var i = 0; i < L - 3; i++)
				res = res.concat(spline(P[i], P[i + 1], P[i + 2], P[i + 3], tension, segs));

			if (close){
				return res.concat(
					spline(P[L - 3], P[L - 2], P[L - 1], P[0], tension, segs)).concat(
					spline(P[L - 2], P[L - 1], P[0], P[1], tension, segs));
			}
			return res.concat(
				spline(P[L - 3], P[L - 2], P[L - 1], P[L - 1], tension, segs)).concat(
				P[L - 1].concat());
		};
	}
	spline2d = makefunc(spline2d_segment);
	spline3d = makefunc(spline3d_segment);
	spline4d = makefunc(spline4d_segment);
})();

if (typeof module !== 'undefined' && module.exports){
	// inside node.js, so export functions into global namespace
	global.spline2d_segment = spline2d_segment;
	global.spline3d_segment = spline3d_segment;
	global.spline4d_segment = spline4d_segment;
	global.spline2d = spline2d;
	global.spline3d = spline3d;
	global.spline4d = spline4d;
}

File: spline.c
// (c) Copyright 2017, Sean Connelly (@voidqk), http://syntheti.cc
// MIT License
// Project Home: https://github.com/voidqk/spline

#include "spline.h"
#include <math.h>

static inline float tb2d(float ax, float ay, float bx, float by, float tension){
	float dx = bx - ax;
	float dy = by - ay;
	float res = powf(dx * dx + dy * dy, tension * 0.5f);
	return res < 0.0001f ? 0.0001f : res;
}

void spline2d_segment(vec2 p0, vec2 p1, vec2 p2, vec2 p3, float tension, int segs, vec2 *out){
	float
		p0x = p0.v[0], p0y = p0.v[1],
		p1x = p1.v[0], p1y = p1.v[1],
		p2x = p2.v[0], p2y = p2.v[1],
		p3x = p3.v[0], p3y = p3.v[1];

	float
		t1 = tb2d(p0x, p0y, p1x, p1y, tension),
		t2 = tb2d(p1x, p1y, p2x, p2y, tension) + t1,
		t3 = tb2d(p2x, p2y, p3x, p3y, tension) + t2;

	float
		dt21 = t2 - t1,
		dt31 = t3 - t1,
		dt32 = t3 - t2;

	for (int s = 0; s < segs; s++){
		float t = dt21 * s / segs + t1;

		float
			dt1 = t1 - t,
			dt2 = t2 - t,
			dt3 = t3 - t;

		float
			A1x = (dt1 * p0x + t * p1x) / t1,
			A1y = (dt1 * p0y + t * p1y) / t1;

		float
			A2x = (dt2 * p1x - dt1 * p2x) / dt21,
			A2y = (dt2 * p1y - dt1 * p2y) / dt21;

		float
			A3x = (dt3 * p2x - dt2 * p3x) / dt32,
			A3y = (dt3 * p2y - dt2 * p3y) / dt32;

		float
			B1x = (dt2 * A1x + t * A2x) / t2,
			B1y = (dt2 * A1y + t * A2y) / t2;

		float
			B2x = (dt3 * A2x - dt1 * A3x) / dt31,
			B2y = (dt3 * A2y - dt1 * A3y) / dt31;

		out[s] = (vec2){
			(dt2 * B1x - dt1 * B2x) / dt21,
			(dt2 * B1y - dt1 * B2y) / dt21
		};
	}
}

static inline float tb3d(float ax, float ay, float az, float bx, float by, float bz, float tension){
	float dx = bx - ax;
	float dy = by - ay;
	float dz = bz - az;
	float res = powf(dx * dx + dy * dy + dz * dz, tension * 0.5f);
	return res < 0.0001f ? 0.0001f : res;
}

void spline3d_segment(vec3 p0, vec3 p1, vec3 p2, vec3 p3, float tension, int segs, vec3 *out){
	float
		p0x = p0.v[0], p0y = p0.v[1], p0z = p0.v[2],
		p1x = p1.v[0], p1y = p1.v[1], p1z = p1.v[2],
		p2x = p2.v[0], p2y = p2.v[1], p2z = p2.v[2],
		p3x = p3.v[0], p3y = p3.v[1], p3z = p3.v[2];

	float
		t1 = tb3d(p0x, p0y, p0z, p1x, p1y, p1z, tension),
		t2 = tb3d(p1x, p1y, p1z, p2x, p2y, p2z, tension) + t1,
		t3 = tb3d(p2x, p2y, p2z, p3x, p3y, p3z, tension) + t2;

	float
		dt21 = t2 - t1,
		dt31 = t3 - t1,
		dt32 = t3 - t2;

	for (int s = 0; s < segs; s++){
		float t = dt21 * s / segs + t1;

		float
			dt1 = t1 - t,
			dt2 = t2 - t,
			dt3 = t3 - t;

		float
			A1x = (dt1 * p0x + t * p1x) / t1,
			A1y = (dt1 * p0y + t * p1y) / t1,
			A1z = (dt1 * p0z + t * p1z) / t1;

		float
			A2x = (dt2 * p1x - dt1 * p2x) / dt21,
			A2y = (dt2 * p1y - dt1 * p2y) / dt21,
			A2z = (dt2 * p1z - dt1 * p2z) / dt21;

		float
			A3x = (dt3 * p2x - dt2 * p3x) / dt32,
			A3y = (dt3 * p2y - dt2 * p3y) / dt32,
			A3z = (dt3 * p2z - dt2 * p3z) / dt32;

		float
			B1x = (dt2 * A1x + t * A2x) / t2,
			B1y = (dt2 * A1y + t * A2y) / t2,
			B1z = (dt2 * A1z + t * A2z) / t2;

		float
			B2x = (dt3 * A2x - dt1 * A3x) / dt31,
			B2y = (dt3 * A2y - dt1 * A3y) / dt31,
			B2z = (dt3 * A2z - dt1 * A3z) / dt31;

		out[s] = (vec3){
			(dt2 * B1x - dt1 * B2x) / dt21,
			(dt2 * B1y - dt1 * B2y) / dt21,
			(dt2 * B1z - dt1 * B2z) / dt21
		};
	}
}

static inline float tb4d(float ax, float ay, float az, float aw, float bx, float by, float bz, float bw,
	float tension){
	float dx = bx - ax;
	float dy = by - ay;
	float dz = bz - az;
	float dw = bw - aw;
	float res = powf(dx * dx + dy * dy + dz * dz + dw * dw, tension * 0.5f);
	return res < 0.0001f ? 0.0001f : res;
}

void spline4d_segment(vec4 p0, vec4 p1, vec4 p2, vec4 p3, float tension, int segs, vec4 *out){
	float
		p0x = p0.v[0], p0y = p0.v[1], p0z = p0.v[2], p0w = p0.v[3],
		p1x = p1.v[0], p1y = p1.v[1], p1z = p1.v[2], p1w = p1.v[3],
		p2x = p2.v[0], p2y = p2.v[1], p2z = p2.v[2], p2w = p2.v[3],
		p3x = p3.v[0], p3y = p3.v[1], p3z = p3.v[2], p3w = p3.v[3];

	float
		t1 = tb4d(p0x, p0y, p0z, p0w, p1x, p1y, p1z, p1w, tension),
		t2 = tb4d(p1x, p1y, p1z, p1w, p2x, p2y, p2z, p2w, tension) + t1,
		t3 = tb4d(p2x, p2y, p2z, p2w, p3x, p3y, p3z, p3w, tension) + t2;

	float
		dt21 = t2 - t1,
		dt31 = t3 - t1,
		dt32 = t3 - t2;

	for (int s = 0; s < segs; s++){
		float t = dt21 * s / segs + t1;

		float
			dt1 = t1 - t,
			dt2 = t2 - t,
			dt3 = t3 - t;

		float
			A1x = (dt1 * p0x + t * p1x) / t1,
			A1y = (dt1 * p0y + t * p1y) / t1,
			A1z = (dt1 * p0z + t * p1z) / t1,
			A1w = (dt1 * p0w + t * p1w) / t1;

		float
			A2x = (dt2 * p1x - dt1 * p2x) / dt21,
			A2y = (dt2 * p1y - dt1 * p2y) / dt21,
			A2z = (dt2 * p1z - dt1 * p2z) / dt21,
			A2w = (dt2 * p1w - dt1 * p2w) / dt21;

		float
			A3x = (dt3 * p2x - dt2 * p3x) / dt32,
			A3y = (dt3 * p2y - dt2 * p3y) / dt32,
			A3z = (dt3 * p2z - dt2 * p3z) / dt32,
			A3w = (dt3 * p2w - dt2 * p3w) / dt32;

		float
			B1x = (dt2 * A1x + t * A2x) / t2,
			B1y = (dt2 * A1y + t * A2y) / t2,
			B1z = (dt2 * A1z + t * A2z) / t2,
			B1w = (dt2 * A1w + t * A2w) / t2;

		float
			B2x = (dt3 * A2x - dt1 * A3x) / dt31,
			B2y = (dt3 * A2y - dt1 * A3y) / dt31,
			B2z = (dt3 * A2z - dt1 * A3z) / dt31,
			B2w = (dt3 * A2w - dt1 * A3w) / dt31;

		out[s] = (vec4){
			(dt2 * B1x - dt1 * B2x) / dt21,
			(dt2 * B1y - dt1 * B2y) / dt21,
			(dt2 * B1z - dt1 * B2z) / dt21,
			(dt2 * B1w - dt1 * B2w) / dt21
		};
	}
}

int splinesize(int pointslen, int segs, bool close){
	if (pointslen <= 0)
		return 0;
	else if (pointslen == 1)
		return 1;
	if (close)
		return segs * pointslen;
	return segs * (pointslen - 1) + 1;
}

#define MAKEFUNC(splineXd, vecX, spline)                                                           \
	void splineXd(int ptslen, const vecX *points, float tension, int segs, bool close, vecX *out){ \
		const vecX *P = points;                                                                    \
		int L = ptslen;                                                                            \
		/* */                                                                                      \
		if (L <= 0)                                                                                \
			return;                                                                                \
		else if (L == 1){                                                                          \
			out[0] = P[0];                                                                         \
			return;                                                                                \
		}                                                                                          \
		else if (L == 2){                                                                          \
			if (close){                                                                            \
				spline(P[1], P[0], P[1], P[0], tension, segs, &out[0]);                            \
				spline(P[0], P[1], P[0], P[1], tension, segs, &out[segs]);                         \
			}                                                                                      \
			else{                                                                                  \
				spline(P[0], P[0], P[1], P[1], tension, segs, &out[0]);                            \
				out[segs] = P[1];                                                                  \
			}                                                                                      \
			return;                                                                                \
		}                                                                                          \
		/* */                                                                                      \
		if (close){                                                                                \
			spline(P[L - 1], P[0], P[1], P[2], tension, segs, &out[0]);                            \
			if (L == 3){                                                                           \
				spline(P[0], P[1], P[2], P[0], tension, segs, &out[segs]);                         \
				spline(P[1], P[2], P[0], P[1], tension, segs, &out[segs * 2]);                     \
				return;                                                                            \
			}                                                                                      \
		}                                                                                          \
		else{                                                                                      \
			spline(P[0], P[0], P[1], P[2], tension, segs, &out[0]);                                \
			if (L == 3){                                                                           \
				spline(P[0], P[1], P[2], P[2], tension, segs, &out[segs]);                         \
				out[segs * 2] = P[2];                                                              \
				return;                                                                            \
			}                                                                                      \
		}                                                                                          \
		/* */                                                                                      \
		for (int i = 0; i < L - 3; i++)                                                            \
			spline(P[i], P[i + 1], P[i + 2], P[i + 3], tension, segs, &out[(i + 1) * segs]);       \
		/* */                                                                                      \
		if (close){                                                                                \
			spline(P[L - 3], P[L - 2], P[L - 1], P[0], tension, segs, &out[(L - 2) * segs]);       \
			spline(P[L - 2], P[L - 1], P[0], P[1], tension, segs, &out[(L - 1) * segs]);           \
			return;                                                                                \
		}                                                                                          \
		spline(P[L - 3], P[L - 2], P[L - 1], P[L - 1], tension, segs, &out[(L - 2) * segs]);       \
		out[(L - 1) * segs] = P[L - 1];                                                            \
	}

MAKEFUNC(spline2d, vec2, spline2d_segment)
MAKEFUNC(spline3d, vec3, spline3d_segment)
MAKEFUNC(spline4d, vec4, spline4d_segment)

#undef MAKEFUNC
